diff --git a/src/lib/ndpi_main.c b/src/lib/ndpi_main.c
index 787fe60..68efb78 100644
--- a/src/lib/ndpi_main.c
+++ b/src/lib/ndpi_main.c
@@ -26,6 +26,7 @@
 #ifndef __KERNEL__
 #include <stdlib.h>
 #include <errno.h>
+#include <time.h>
 #endif
 
 #include "ndpi_main.h"
@@ -34,7 +35,6 @@
 
 #include "ahocorasick.h"
 
-#include <time.h>
 
 #ifdef __KERNEL__
 #define printf printk
@@ -5069,4 +5069,4 @@ int ndpi_match_string_subprotocol(struct ndpi_detection_module_struct *ndpi_stru
 }
 
 
-	
\ No newline at end of file
+	
