--- a/Makefile	2012-10-12 10:15:22.000000000 +0200
+++ b/Makefile	2012-10-12 13:14:01.291197339 +0200
@@ -179,7 +179,7 @@ ${BUILDDIR}/src/ffdecsa/ffdecsa_sse2.o :
 
 # File bundles
 SRCS-${CONFIG_BUNDLE}     += bundle.c
-BUNDLES-yes               += docs/html docs/docresources src/webui/static
+BUNDLES-yes               += src/webui/static
 BUNDLES-yes               += data/conf
 BUNDLES-${CONFIG_DVBSCAN} += data/dvb-scan
 BUNDLES                    = $(BUNDLES-yes)
