Index: vtun-3.0.3/configure.ac
===================================================================
--- vtun-3.0.3.orig/configure.ac	2016-01-11 21:06:10.025878121 +0800
+++ vtun-3.0.3/configure.ac	2016-01-11 21:06:58.305877257 +0800
@@ -136,7 +136,7 @@
    AC_CHECKING( for LZO Library and Header files ... )
    AC_SEARCH_HEADERS(lzo_asm.h,
 	$LZO_HDR_DIR /usr/include/lzo "" /usr/local/include,
-	LZOCHK="lzo2 lzo",
+	LZOCHK="lzo2",
    AC_SEARCH_HEADERS(lzo1x.h, 
 	$LZO_HDR_DIR /usr/include/lzo "" /usr/local/include,
 		LZOCHK="lzo",
@@ -172,10 +172,9 @@
                   [AC_DEFINE(HAVE_SODIUM, [1], [Define to 1 if you have Sodium])],
                   AC_ERROR([libsodium with crypto_aead_aes256gcm_encrypt not found]))
    AC_SEARCH_HEADERS(sodium.h,
-    $SSL_HDR_DIR /usr/include /usr/local/include /opt/include, 
-    ,
-    AC_MSG_ERROR( Sodium headers not found. )
-   )
+   		$SODIUM_HDR_DIR /usr/include /usr/local/include /opt/include, 
+    		,
+    		AC_MSG_ERROR( Sodium headers not found. ))
 fi
 
 if test "$NATHACK" = "yes"; then
@@ -202,9 +201,6 @@
 
 AC_MSG_RESULT()
 
-CFLAGS="$CFLAGS -fstack-protector-all -fno-strict-overflow -fno-strict-aliasing"
-CPPFLAGS="$CPPFLAGS -D_FORTIFY_SOURCE=2"
-
 AC_CHECK_FUNCS([getpt grantpt unlockpt ptsname])
 
 OS_REL=`uname -r | tr -d '[A-Za-z\-\_\.]'`
