diff -uNr a/auto/cc/name b/auto/cc/name
--- a/auto/cc/name	2013-11-22 15:47:33.000000000 +0800
+++ b/auto/cc/name	2014-11-11 16:59:06.469301606 +0800
@@ -18,7 +18,7 @@
         echo
         echo $0: error: C compiler $CC is not found
         echo
-        exit 1
+#       exit 1
     fi
 
 fi
diff -uNr a/auto/types/sizeof b/auto/types/sizeof
--- a/auto/types/sizeof	2013-11-22 15:47:33.000000000 +0800
+++ b/auto/types/sizeof	2014-11-11 16:56:31.853302647 +0800
@@ -25,8 +25,13 @@
 $NGX_INCLUDE_INTTYPES_H
 $NGX_INCLUDE_AUTO_CONFIG_H
 
+char object_code_block[] = {
+	'\n', 'e', '4', 'V', 'A',
+	'0', 'x', ('0' + sizeof($ngx_type)),
+	'Y', '3', 'p', 'M', '\n'
+};
+
 int main() {
-    printf("%d", (int) sizeof($ngx_type));
     return 0;
 }
 
@@ -40,7 +45,7 @@
 
 
 if [ -x $NGX_AUTOTEST ]; then
-    ngx_size=`$NGX_AUTOTEST`
+    ngx_size=`sed -ne 's/^e4VA0x\(.\)Y3pM$/\1/p' < $NGX_AUTOTEST`
     echo " $ngx_size bytes"
 fi
 
diff -uNr a/auto/unix b/auto/unix
--- a/auto/unix	2013-11-22 15:47:33.000000000 +0800
+++ b/auto/unix	2014-11-11 17:20:43.021292864 +0800
@@ -715,7 +715,7 @@
 
 ngx_feature="System V shared memory"
 ngx_feature_name="NGX_HAVE_SYSVSHM"
-ngx_feature_run=yes
+ngx_feature_run=no
 ngx_feature_incs="#include <sys/ipc.h>
                   #include <sys/shm.h>"
 ngx_feature_path=
